Modify translation map

knelli2 opened the pull request:
## Proposed changes
Previously, each component of a translation had to be stored as a separate 1d translation map. If you wanted a 2d or 3d translation map, you'd need to do a product of maps. This simplifies the interface by templating the translation map on the dimension.

This was prompted by control system edits. For translation control, we only want one control system controlling one function of time that represents a translation in 3d. With the old implementation of the translation, we would have needed three separate control systems controlling three separate functions of time. That would not have been ideal.

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions
- If you have any Translation maps here are the new versions:
```
Old  ->  New
Translation  ->  Translation<1>
ProductOf2Maps<Translation, Translation>  ->  Translation<2>
ProductOf3Maps<Translation, Translation, Translation>  ->  Translation<3>
```
- The UniformTranslation time dependence for a domain creator now takes as its last argument a `std::string` instead of a `std::array<std::string, Dim>` because we no longer need `Dim` number of functions of time to represent a translation map. We only need 1. Thus, in an input file, just a single string is needed for the name of the function of time, not a list.
- The option for the name of the function of time has been renamed from `FunctionOfTimeNames` to `FunctionOfTimeName`. Thus, with the above change, a typical UniformTranslation time dependence in 1 dimension would look like this in an input file:
```
UniformTranslation:
    InitialTime: 0.0
    InitialExpirationDeltaT: 2.0
    Velocity: [0.5]
    FunctionOfTimeName: "Translation1D"
```
and in 3 dimensions...
```
UniformTranslation:
    InitialTime: 0.0
    InitialExpirationDeltaT: 2.0
    Velocity: [0.5, 0.4, 0.3]
    FunctionOfTimeName: "Translation3D"
```

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

wthrowe commented on tests/Unit/Domain/Test_TagsTimeDependent.cpp:
```diff
@@ -128,11 +119,10 @@ void test() noexcept {
 
   MAKE_GENERATOR(gen);
-  const std::array<double, 3> velocity{{1.2, 0.2, -8.9}};
+  const DataVector velocity{Dim, 1.2};
```
Why the change in value here?

knelli2 replied:
Below, on old lines 152-166, each component of the `std::array` was used to construct a `DataVector` of length 1 for each function of time (3 in total). Since we only need 1 function of time now, I just made the original `std::array` a `DataVector` of appropriate length (depending on the Dim).

Same for the other comments.

wthrowe replied:
But why the change from {1.2, 0.2, -8.9} to {1.2, 1.2, 1.2}?

wthrowe replied:
Oh, right, it needs to be the right size now.  Never mind.

knelli2 replied:
Oh the values are arbitrary so instead of adding `if` statements to construct a separate `DataVector` for each Dim I just did this since it was simpler.

wthrowe commented on tests/Unit/Evolution/Initialization/Test_DgDomain.cpp:
```diff
@@ -203,22 +176,19 @@ void test(const Spectral::Quadrature quadrature) noexcept {
       make_array<Dim>(0_st)};
   const size_t num_pts = pow<Dim>(4_st);
-  const std::array<double, 3> velocity{{1.2, 0.2, -8.9}};
+  const DataVector velocity{Dim, 3.6};
```
Again, why this change?

wthrowe commented on tests/Unit/Evolution/Test_TagsDomain.cpp:
```diff
@@ -114,14 +73,13 @@ void test() noexcept {
       evolution::domain::Tags::DivMeshVelocityCompute<Dim>>;
 
-  const std::array<double, 3> velocity{{1.2, 0.2, -8.9}};
+  const DataVector velocity{Dim, -4.3};
```
Again.

wthrowe reviewed (changes requested):
In the future, if you are going to reformat a lot of code that you are not otherwise touching, please do it in a separate commit.

knelli2 commented:
Oops, guess I got a little carried away with the auto-formatting 😅 . I can split it up here if you want

wthrowe reviewed (dismissed):
Up to you on splitting the commit.  This is long enough that I want another reviewer, and they may ask for it to be done, though.

knelli2 committed:
Modify translation map

Previously, each component of a translation had to be stored as a separate
1d translation map. If you wanted a 2d or 3d translation map, you'd need
to do a product of maps. This simplifies the interface by templating
the translation map on the dimension.

knelli2 commented:
Ok I decided to just undo the formatting changes that weren't relevant to this PR. I think I got everything. Let me know if I didn't.

nilsdeppe commented on src/Domain/CoordinateMaps/TimeDependent/Translation.cpp:
```diff
@@ -62,58 +87,94 @@ std::array<tt::remove_cvref_wrap_t<T>, 1> Translation::frame_velocity(
+  std::array<tt::remove_cvref_wrap_t<T>, Dim> result{};
+  const auto function_of_time_and_deriv =
+      functions_of_time.at(f_of_t_name_)->func_and_deriv(time);
+  const DataVector velocity = function_of_time_and_deriv[1];
```
I thought I explicitly asked when this was discussed if the velocity would be `std::array<double, Dim>` and not a `DataVector` to which I remember being told "yes". I guess this isn't true? Am I misremembering what I was told could be a `std::array<double>`? I don't have a change request, I'm just confused that this disagrees with what I remember. I guess previously it was a `DataVector` as well instead of a `double` or `std::arary<double, 1>`?

`const DataVector` -> `const DataVector&` should work, no?

knelli2 replied:
Hm I guess I'm confused by your confusion. Nothing about the outward interface of this function has changed. The source coordinates input type is still either a `std::array<double>` or `std::array<DataVector>` and the return type is still a `std::array<double>` or `std::array<DataVector>` corresponding to the input type (see line 90). Nothing has changed in that regard.

The velocity defined on line 93 is just a temporary. The `make_with_value` on line 100 constructs the appropriate return type (`double` or `DataVector`) for each element of the resulting `std::array`. Does that answer your question?

And yes, `const DataVector&` works, I'll put that in.

nilsdeppe replied:
Yea, I was probably just confused at the time of the discussion I'm remembering :) The code here is what I expect based on what I wrote 😄 

:+1:

nilsdeppe commented on tests/Unit/Domain/Creators/TimeDependence/Test_UniformTranslation.cpp:
```diff
@@ -31,54 +31,24 @@
 namespace domain::creators::time_dependence {
 
 namespace {
```
is now possible to remove the `ProductMaps.?pp` includes?

knelli2 replied:
Oops, yup forgot to take out all these includes.

nilsdeppe commented on tests/Unit/Domain/Test_Block.cpp:
```diff
@@ -99,52 +99,19 @@ void test_block_time_independent() {
   test_move_semantics(std::move(original_block), block_copy);
 }
 
```
do you need the time-dependent product maps include still?

nilsdeppe commented on tests/Unit/Domain/Test_TagsTimeDependent.cpp:
```diff
@@ -58,13 +58,9 @@ void test_tags() noexcept {
       "div(MeshVelocity)");
 }
 
```
remove include of time-dependent product maps?

nilsdeppe commented on tests/Unit/Evolution/Initialization/Test_DgDomain.cpp:
```diff
@@ -39,12 +39,9 @@
 #include "Utilities/TMPL.hpp"
 
 namespace {
```
remove include of time-dependent product maps?

nilsdeppe commented on tests/Unit/Evolution/Test_TagsDomain.cpp:
```diff
@@ -43,57 +42,17 @@ void test_tags() noexcept {
       "div(MeshVelocity)");
 }
 
```
remove include of time-dependent product maps?

nilsdeppe reviewed (changes requested):
Overall looks good. Just a few comments. Please squash directly :)

nilsdeppe reviewed (approved):
@wthrowe I'm happy for this to be merged
